From: Nikhil Benesch <nikhil.benesch@gmail.com>
Date: Mon, 29 Feb 2016 11:19:03 -0800
Subject: handle new files

---
 bus/dev_emufs.c | 8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

diff --git a/bus/dev_emufs.c b/bus/dev_emufs.c
index 44b3a2f..9ae11a8 100644
--- a/bus/dev_emufs.c
+++ b/bus/dev_emufs.c
@@ -302,6 +302,7 @@ emufs_open(struct emufs_data *ed, int flags)
 	int curdir;
 	struct stat sbuf, sbuf2;
 	int isdir;
+	int create = 0;
 
 	if (ed->ed_iolen >= EMU_BUF_SIZE) {
 		return EMU_RES_BADSIZE;
@@ -327,6 +328,9 @@ emufs_open(struct emufs_data *ed, int flags)
 		/* creating; ok if it doesn't exist, and it's not a dir */
 		flags |= O_RDWR;
 		isdir = 0;
+		sbuf.st_dev = 0;
+		sbuf.st_ino = 0;
+		create = 1;
 	}
 	else {
 		isdir = S_ISDIR(sbuf.st_mode)!=0;
@@ -370,8 +374,8 @@ emufs_open(struct emufs_data *ed, int flags)
 		 * when we finally get code to prohibit going outside
 		 * the root dir.
 		 */
-		if (sbuf2.st_dev != sbuf.st_dev ||
-		    sbuf2.st_ino != sbuf.st_ino) {
+		if (!create && (sbuf2.st_dev != sbuf.st_dev ||
+				sbuf2.st_ino != sbuf.st_ino)) {
 			close(ed->ed_handles[handle].eh_fd);
 			ed->ed_handles[handle].eh_fd = -1;
 			goto retry;
